opengever/maintenance/scripts/fix_non_persisted_values.py

```diff
@@ -0,0 +1,952 @@
+        """
+        return self._get_volatile_value(obj, field)
+
+    def get_custody_period_value(self, obj, field):
+        """Get value for custody_period field.
+        """
+        return self._get_volatile_value(obj, field)
+
+    def get_retention_period_value(self, obj, field):
+        """Get value for retention_period field.
+        """
+        return self._get_volatile_value(obj, field)
+
+    def get_privacy_layer_value(self, obj, field):
+        """Get value for privacy_layer field.
+        """
+        return self._get_volatile_value(obj, field)
+
+    handlers = {
+        ('IDocumentMetadata', 'preserved_as_paper'): get_preserved_as_paper_value,  # noqa
+        ('IDocumentMetadata', 'document_date'): get_document_date_value,
+        ('start'): get_dossier_start_value,
+        ('IProtectDossier', 'dossier_manager'): get_dossier_manager_value,
+        ('ITask', 'deadline'): get_deadline_value,
+        ('ILifeCycle', 'custody_period'): get_custody_period_value,
+        ('ILifeCycle', 'retention_period'): get_retention_period_value,
+        ('IClassification', 'public_trial'): get_public_trial_value,
+        ('IClassification', 'classification'): get_classification_value,
+        ('IClassification', 'privacy_layer'): get_privacy_layer_value,
+    }
+
+
+class Reindexer(object):
+
+    def __init__(self, fixer):
+        self.catalog = fixer.catalog
+        self.stats = fixer.stats
+
+        self._metadata_names = None
+        self._index_names = None
+
+    @property
+    def metadata_names(self):
+        """List of metadata column names (memoized).
+        """
+        if self._metadata_names is None:
+            self._metadata_names = self.catalog._catalog.schema.keys()
+        return self._metadata_names
+
+    @property
+    def index_names(self):
+        """List of index names (memoized).
+        """
+        # we ignore indexed attrs here
+        if self._index_names is None:
+            self._index_names = self.catalog.indexes()
+        return self._index_names
+
+    def reindex_if_necessary(self, obj, fixed_fields, brain):
+        """Reindex indexes and metadata for the given object if needed.
+        """
+        update_metadata = self.needs_metadata_update(
+            obj, fixed_fields, brain)
+
+        self.stats['update_metadata'][update_metadata] += 1
+
+        idxs_needing_reindex = self.get_idxs_needing_reindex(
+            obj, fixed_fields, brain)
+
+        if update_metadata or idxs_needing_reindex:
+            # If idxs == [] the catalog defaults to *all* indexes, so we
+            # supply it a cheap index to trick it into only rebuilding metadata
+            if not idxs_needing_reindex:
+                idxs_needing_reindex = ['getId']
+
+            sys.stderr.write(
+                "Reindexing %s (update_metadata=%r, idxs=%r)\n" % (
+                    obj, update_metadata, idxs_needing_reindex))
+            self.catalog.reindexObject(obj, idxs=idxs_needing_reindex,
```

Comment:
For SOLR deployments we would need to do a manual solr-reindex here, because solr reindex is not triggered when using `catalog.reindexObject`.

Reply:
@phgross added support for Solr indexing. This currently works for deployments that have Solr enabled and ones that don't - but for older GEVER releases that don't even have `ftw.solr` in their eggs there still needs to be conditional imports. But I'll adapt the script once we get into that situation.